### PR TITLE
#42: UDF support

### LINKED ISSUE #42: Add support for building UDF Scripts

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #42

### CODE REVIEW

File: src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java [hunk 1]
```diff
@@ -30,13 +30,15 @@
...
+import com.exasol.dbbuilder.dialects.exasol.udf.UdfScript;
 
 @Tag("integration")
 @Testcontainers
 // [itest->dsn~exasol-object-factory~1]
 class ExasolDatabaseObjectCreationAndDeletionIT extends AbstractDatabaseObjectCreationAndDeletionIT {
     @Container
-    private static final ExasolContainer<? extends ExasolContainer<?>> container = new ExasolContainer<>();
+    private static final ExasolContainer<? extends ExasolContainer<?>> container = new ExasolContainer<>()
+            .withReuse(true);
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:41]: why is this needed now?

File: src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java [hunk 2]
```diff
@@ -47,9 +49,12 @@ protected Connection getAdminConnection() throws SQLException {
         return container.createConnection("");
     }
 
-    @Override
-    protected DatabaseObjectFactory getDatabaseObjectFactory() throws SQLException {
-        return new ExasolObjectFactory(container.createConnection(""));
+    private static String getTableSysName(final DatabaseObject object) {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:52]: as this method seems to be used only in `ExistsInDatabaseMatcher`,  I would suggest to move it there

**Review by chiaradiamarcelo — CHANGES_REQUESTED** [review]
Looks good! left some comments

File: doc/changes/changes_2.0.0.md [hunk 3]
```diff
@@ -7,3 +7,18 @@ Code name:
 * #40: Refactored debugger connection to global JVM options
    This change caused an API change for adding a debugger. 
    See the [user guide](../user_guide/user_guide.md). 
+* #42: UDF support
+    For supporting UDFs we unified the API for of Script, Udf and Adapter scripts.
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply doc/changes/changes_2.0.0.md:11]:
```suggestion
    For supporting UDFs we unified the API for Script, UDF and Adapter scripts.
```

File: src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java (hunk 1 above)
Comment (jakobbraun) [reply src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:41]: it makes allows you to make tests faster. I just added it because it is possible and a new feature

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/udf/UdfScript.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        @Override
+        protected Builder getSelf() {
+            return this;
+        }
+    }
+
+    /**
+     * This class represents the {@code EMITS(...)} {@link UdfReturnType}.
+     */
+    public static class UdfReturnTypeDynamicEmits implements UdfReturnType {
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/udf/UdfScript.java:220]: i forgot it...

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java [hunk 5]
```diff
@@ -30,4 +40,42 @@ public boolean hasParent() {
...
+            this.parentSchema = parentSchema;
+            this.name = name;
+        }
+
+        /**
+         * Validate the fields. Call this in {@code build()}.
+         */
+        protected void validate() {
+            requireNotNull(this.parentSchema, "parentSchema");
+            requireNotNull(this.name, "name");
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java:65]: 1. Objects.requireNonNull(object, message) throws NullPinter exception while I prefer illegalState here
2. by this custom requireNotNull I avoid duplication in messages and make the uniform

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java [hunk 6]
```diff
@@ -30,4 +40,42 @@ public boolean hasParent() {
...
+        protected Builder(final Schema parentSchema, final Identifier name) {
+            this.parentSchema = parentSchema;
+            this.name = name;
+        }
+
+        /**
+         * Validate the fields. Call this in {@code build()}.
+         */
+        protected void validate() {
+            requireNotNull(this.parentSchema, "parentSchema");
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java:64]: As far as I know we usualy do it like that: The builder checks if all required fileds were set before building.
@redcatbear what's your oppinion?

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java [hunk 7]
```diff
@@ -136,6 +148,65 @@ public void drop(final Script script) {
...
+        sqlBuilder.append("\n/");
+        writeToObject(udfScript, sqlBuilder.toString());
+    }
+
+    // [impl->dsn~dropping-udfs~1]
+    public void drop(final UdfScript udfScript) {
+        writeToObject(udfScript, "DROP SCRIPT " + udfScript.getFullyQualifiedName());
+    }
+
+    private static class UdfReturnTypeWriter implements UdfReturnTypeVisitor {
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java:183]: I think the visitor is useful if we want to use the classes from outside.

[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java:183]: For that reason I think it is anlways a good idea to offer a visitor interface...

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolSchema.java [hunk 8]
```diff
@@ -37,17 +38,28 @@ public ExasolSchema(final ExasolImmediateDatabaseObjectWriter writer, final Stri
...
-        return AdapterScript.builder().writer(this.writer).parentSchema(this).name(name).language(language)
-                .content(content).build();
+        return AdapterScript.builder(this.writer, this, ExasolIdentifier.of(name)).language(language).content(content)
+                .build();
     }
 
     /**
      * Create a builder for an adapter script.
      *
+     * @param name name of the adapter script
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolSchema.java:48]: Well an identifier is the class that models object names like MY_SCHEMA."test"
But in this case it's still the adapter name. Even so it's a qualified identifier

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/AdapterScript.java [hunk 9]
```diff
@@ -116,71 +74,27 @@ public Builder language(final Language language) {
...
          * 
          * @return built {@link AdapterScript}.
          */
         public AdapterScript build() {
-            requireNotNull(this.writer, "writer");
-            requireNotNull(this.parentSchema, "parentSchema");
-            requireNotNull(this.name, "name");
+            validate();
+            final AdapterScript adapterScript = new AdapterScript(this);
+            getWriter().write(adapterScript);
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/AdapterScript.java:85]: I guess the reason why @redcatbear didi it like that is taht the purpose of this tool lib is to create test-setups as easy as possible.
If you think we should change it, open a ticket, since it affects the whole design and is not realy part of this PR.

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java:62]: I want validation to be done before the object is constructed. Hence the check in the builder.

If the object is immutable, the result will be and remain valid.

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/AdapterScript.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/AdapterScript.java:85]: The main point was compact code. It was a conscious design decision that creating a POJO is mirrored in the database. I would not do that in production code, but here it makes the tests a lot more compact.

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java:62]: @redcatbear  yes, I understand. The issue with this is that Builder is not thread safe, so some other thread might be changing the values while constructing the resulting object, and you can end up with an invalid instance.

I'm not fully aware of how we are using this and if there's a possibility for such scenario.

[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchemaChild.java:62]: Sharing the builder reference between threads is obviously not a good idea. The only other option would be validation in the constructor and I am generally against throwing exceptions from constructors.

**Review by chiaradiamarcelo — APPROVED** [review]

**Review by chiaradiamarcelo — APPROVED** [review]
